core/src/main/codegen/templates/Parser.jj

```diff
@@ -2092,13 +2096,32 @@ SqlNode JoinTable(SqlNode e) :
     //
     // We allow CROSS JOIN (joinType = CROSS_JOIN) to have a join condition,
     // even though that is not valid SQL; the validator will catch it.
-    LOOKAHEAD(3)
+    LOOKAHEAD(4)
     natural = Natural()
     joinType = JoinType()
     e2 = TableRef1(ExprContext.ACCEPT_QUERY_OR_JOIN)
     (
+        [ <MATCH_CONDITION> matchCondition = Expression(ExprContext.ACCEPT_SUB_QUERY) ]
         <ON> { on = JoinConditionType.ON.symbol(getPos()); }
```

Comment:
@mihaibudiu this means that the `ON` clause is mandatory for `ASOF` joins as well, is that intentional? Snowflake's `ASOF JOIN` only requires the `MATCH_CONDITION`, with `ON` / `USING` being optional - https://docs.snowflake.com/en/sql-reference/constructs/asof-join#syntax.

Reply:
I guess that it is mandatory with this implementation
And you can't just use 'true' either, since it requires an equijoin

Reply:
> And you can't just use 'true' either, since it requires an equijoin

Actually, this does seem to work because the `ConjunctionOfEqualities` class in `SqlValidatorImpl` doesn't override the literal visit method of `SqlShuttle` so `ConjunctionOfEqualities.illegal` retains its default value of false for a join condition like `ON true`.